Alt-branch-1 ready to be merged into main

Fernando-Iscar opened the pull request:
Hi guys,

As discussed in the last meeting, we will continue working with unit testing in the new branch `test-development-branch`.

Thanks!

@DiSanchz

DiSanchz committed:
test

DiSanchz committed:
test_

DiSanchz committed:
gitignore+pyproject.toml

DiSanchz committed:
Merge branch 'alt_branch_1' of https://github.com/snpe-reputation-systems/snpe-reputation-systems-structure into alt_branch_1

DiSanchz committed:
environment.yml

DiSanchz committed:
dockerfile

DiSanchz committed:
docker-compose

DiSanchz committed:
dockerfile

DiSanchz committed:
structure

DiSanchz committed:
dockerfiles_out

DiSanchz committed:
lint-formatting-draft

DiSanchz committed:
modified by formatting pipeline

DiSanchz committed:
formatting_pipe2

DiSanchz committed:
Merge branch 'alt_branch_1' of https://github.com/snpe-reputation-systems/snpe-reputation-systems-structure into alt_branch_1

DiSanchz committed:
formatting_pipe3

DiSanchz committed:
formatting_pipe3

DiSanchz committed:
modified by formatting pipeline

DiSanchz committed:
formating_test

DiSanchz committed:
Merge branch 'alt_branch_1' of https://github.com/snpe-reputation-systems/snpe-reputation-systems-structure into alt_branch_1

DiSanchz committed:
format_test2

DiSanchz committed:
modified by formatting pipeline

DiSanchz committed:
format_test3

DiSanchz committed:
format_test4

DiSanchz committed:
format_test5

DiSanchz committed:
format_test6

DiSanchz committed:
format_test7

DiSanchz committed:
format_test8

DiSanchz committed:
format_test8

Fernando-Iscar committed:
Create testing-self-hosted-runner.yml

Adapting testing-workflow.yml to a self-hosted runner

DiSanchz committed:
testing_t_workflow

DiSanchz committed:
testing_t_workflow

DiSanchz committed:
ubuntu_lint

DiSanchz committed:
self-hosted

DiSanchz committed:
fix_wf

DiSanchz committed:
modified by formatting pipeline

DiSanchz committed:
twt1

DiSanchz committed:
Merge branch 'alt_branch_1' of https://github.com/snpe-reputation-systems/snpe-reputation-systems-structure into alt_branch_1

DiSanchz committed:
twt2

DiSanchz committed:
twt3

DiSanchz committed:
twt4

DiSanchz committed:
twt5

DiSanchz committed:
modified by formatting pipeline

DiSanchz committed:
twt6

DiSanchz committed:
twt7

DiSanchz committed:
twt8

DiSanchz committed:
modified by formatting pipeline

DiSanchz committed:
new_dockerfile

DiSanchz committed:
fixes

DiSanchz committed:
t_new_wf

DiSanchz committed:
unified_workflow

DiSanchz committed:
fix-unified_wf

DiSanchz committed:
uni_wf

DiSanchz committed:
test_test

DiSanchz committed:
test_test_2

DiSanchz committed:
test_test_3

DiSanchz committed:
utils_added

DiSanchz committed:
test_fix

DiSanchz committed:
test_fix_2

DiSanchz committed:
test_wf

DiSanchz committed:
add_needs_wf

DiSanchz committed:
first_tests

DiSanchz committed:
token_added

DiSanchz committed:
format_pipe_fix

DiSanchz committed:
linter-fix

DiSanchz committed:
format-fix

DiSanchz committed:
format-fix2

DiSanchz committed:
format-fix3

DiSanchz committed:
test_fix2

DiSanchz committed:
test_fix3

DiSanchz committed:
pipe_test

DiSanchz committed:
new_dockerfile

DiSanchz committed:
wf_updated

DiSanchz committed:
test_fix

DiSanchz committed:
labels_added

DiSanchz committed:
unit_tests

DiSanchz committed:
unit_tests

DiSanchz committed:
unit_tests

DiSanchz committed:
unit_test

DiSanchz committed:
unit_test

DiSanchz committed:
unit_test

DiSanchz committed:
unit_test

DiSanchz committed:
unit_test

DiSanchz committed:
unit_test

DiSanchz committed:
unit_test

DiSanchz committed:
unit_tests

DiSanchz committed:
unit_test

DiSanchz committed:
example_tests_deleted

DiSanchz committed:
testing_pipeline

narendramukherjee commented on pyproject.toml:
```diff
@@ -0,0 +1,27 @@
+name = "snpe_reputation_systems"
+version = "0.0.1"
+description = "Dynamic Structural Modeling and Approximate Bayesian Inference of Online Reputation System"
+authors = ["Shrabastee Banerjee", "Narendra Mukherjee", "M. Amin Rahimian"]
```
@Fernando-Iscar @DiSanchz - since this is to do with the authors of the codebase rather than the paper, you guys should add your names as well

Fernando-Iscar replied:
Names added, thanks! 
@narendramukherjee

narendramukherjee commented on snpe_reputation_systems/poetry.lock:
Is the poetry lock file usually committed in a repository as well? Does it change every time you run poetry?

DiSanchz replied:
**Regarding the first question:**

This particular "poetry.lock" file was generated and committed at the time when the functioning of poetry was being tested for the first time. At that point, if I remember correctly, I was still assuming (incorrectly) that poetry would play some role in dependency management in the project.

Having said that, the "poetry.lock" file is indeed usually committed in repositories as a tool to ensure that all collaborators work under the same version of the relevant dependencies. However, as far as I have read, this holds true mostly when at least one of the following conditions is met:

1. The nature of the project makes the inclusion of the lock file particularly interesting or advisable
2. Poetry is the primary tool for dependency management

In our case, none of these are met given that:

1. End users of a Python package interact with it via a package installer, which uses the dependency specifications in pyproject.toml, not "poetry.lock". Therefore, the "poetry.lock" file doesn't influence the package installation for users, making it less relevant to include in a package's repository. Moreover, packages typically aim at being compatible with a range of dependencies, rather than the specific versions "locked" at a given time, which also diminishes the relevance of "poetry.lock" as part of the repository.
2. Poetry is not being used as our primary tool for dependency management.

The "poetry.lock" can also contribute to a project's reproducibility to an extent, however, I understand we are already addressing this with the use of docker containers. Thus the benefits in this sense would be rather limited considering that the yet-to-be-added dockerfile(s) for the new repo contain appropriate instructions on the range of dependency versions within which we expect it to be working.

**Regarding the second question:**

A poetry.lock file is generated whenever the `poetry install` command is run and poetry could not find an existing "poetry.lock" within its context, based on the information on dependencies included in the "pyproject.toml". Moreover, (according to poetry's documentation) it is also updated/generated whenever the project's dependencies are modified through poetry commands (mainly `poetry remove`, `poetry add`, `poetry remove`, which also affect the "pyproject.toml"). So, in short, it does not necessarily change every time the repository is "packaged" as long as the dependency information inside the "pyproject.toml" has not been modified as well.

**As a personal conclusion/opinion**

Whether to commit the "poetry.lock" largely depends on the particular situation. Given the factors considered above and our particular use context, I lean to think that the reasons for not committing it outweigh those others in favor of doing so. If you agree, I would remove it from the branch before merging.

**Sources**

For further insight, there are some other factors at play regarding the tradeoffs of including or not the "poetry.lock" file (which I have considered less relevant). These sources mention some of them, alongside those already discussed above.

https://python-poetry.org/docs/basic-usage/#:~:text=Committing%20your%20poetry.lock%20file%20to%20version%20control,-As%20an%20application&text=Application%20developers%20commit%20poetry.lock,dependencies%20that%20you%20are%20using. [Section: "Committing your poetry.lock file to version control"]

https://github.com/python-poetry/poetry/issues/7488

narendramukherjee replied:
Sounds good @DiSanchz - let's remove the lock file then. Also, at the next meeting, let's briefly go over what's doing dependency management right now and how the library is being packaged

narendramukherjee commented on snpe_reputation_systems/snpe_reputation_systems/utils/functions.py:
```diff
@@ -0,0 +1,111 @@
+# for path in terminal_execute(command):
+#   print(path, end="")
+# https://stackoverflow.com/questions/4417546/constantly-print-subprocess-output-while-process-is-running
+def terminal_execute(cmd: str) -> None:
```
`mypy` has raised an error here as the return type of this function is not actually `None` - it is actually taking a Linux command `cmd`, running it and "yielding" the output of running this command on line 19 (`yield stdout_line`). If you don't do this, the output of a command run from Python in a subprocess, like here, will only be seen after the command is completed. We want to use this to run the starspace embeddings fitting command, which will take a while to run - if you don't have this functionality, you will not know how much Starspace has progressed.

So here the return type is a Python `Generator`, or if only yielding and nothing else, `Iterator` - generators are special Python objects which `yield` things: for eg., the `range` function is also a Generator. Try looking at the type of `range(5)` yourself.

So here, the fix is to have a return type of `Iterator[str]` following the instructions here: https://stackoverflow.com/questions/43658999/what-is-the-return-type-hint-of-a-generator-function
You need to also `from typing import Iterator` of course

Fernando-Iscar replied:
Ok, we implemented these changes 👍 👍

narendramukherjee commented on snpe_reputation_systems/snpe_reputation_systems/utils/embedding_nets.py:
```diff
@@ -0,0 +1,93 @@
+import numpy as np
+import torch
+
```
`mypy` asks to `from typing import List` here. Check log of pipeline

Fernando-Iscar replied:
Ok, now it's working 👍

narendramukherjee reviewed (changes requested):
Suggested the fixes needed for mypy

Fernando-Iscar committed:
Update formatting-testing.yml

If some errors are spotted in formatting workflow, it won't pass into the testing job after setting parameter `continue-on-error` to `false`

Fernando-Iscar committed:
Adding author names to pyproject.toml

DiSanchz commented:
As an additional relevant note on the PR, the last commits by Fernando include a modification of the workflow which now will be marked as failed if any formatting errors are found in the code as we discussed during the last meeting.

Fernando-Iscar committed:
linting w/ black 1

Fernando-Iscar committed:
linting w/ black 2

Fernando-Iscar committed:
linting w/ black 3

Fernando-Iscar committed:
linting w/ black & isort

Fernando-Iscar committed:
linting w/ black & isort2

Fernando-Iscar committed:
linting w/ mypy, black & isort

Fernando-Iscar committed:
linting w/ mypy, black & isort 2

Fernando-Iscar committed:
linting w/ mypy, black & isort 3

Fernando-Iscar commented:
Hi @narendramukherjee @shrabasteebanerjee ,

As you may see below, we solved the issues with the workflow and now all checks are passing. Hence, in that regard, the PR should be ready.

Let us know if you have any further suggestions

narendramukherjee commented:
I have approved the changes, can @shrabasteebanerjee approve these changes as well? We need 2 approvals for merging

@DiSanchz @Fernando-Iscar When the merge button is available when 2 approvals have been obtained, be careful that you select "squash commits while merging" by clicking on the downward triangle next to the merge button. Then the MR makes only 1 merge commit and the commit history is cleaner

DiSanchz commented:
We got it! Let's merge it!
